app/helpers/location_helper.rb

```diff
@@ -75,31 +75,44 @@ def self.sample_dimensions(sample_ids, field_name, samples_count)
     locations = SamplesHelper.samples_by_metadata_field(sample_ids, field_name).count
     locations = locations.map do |loc, count|
       location = loc.is_a?(Array) ? (loc[0] || loc[1]) : loc
-      { value: location, text: truncate_name(location), count: count }
+      parents = loc.is_a?(Array) ? loc[2..-1].compact[0..-2] : []
+      return { value: location, text: truncate_name(location), count: count, parents: parents }
     end
     not_set_count = samples_count - locations.sum { |l| l[:count] }
     if not_set_count > 0
       locations << { value: "not_set", text: "Unknown", count: not_set_count }
     end
-    locations
+    return locations
   end
 
   def self.project_dimensions(sample_ids, field_name)
     # See pattern in ProjectsController dimensions
-    locations = SamplesHelper.samples_by_metadata_field(sample_ids, field_name)
-                             .includes(:sample)
-                             .distinct
-                             .count(:project_id)
+    locations = SamplesHelper
+                .samples_by_metadata_field(sample_ids, field_name)
+                .includes(:sample)
+                .distinct
+                .count(:project_id)
     locations.map do |loc, count|
       location = loc.is_a?(Array) ? (loc[0] || loc[1]) : loc
       { value: location, text: truncate_name(location), count: count }
     end
   end
 
   def self.filter_by_name(samples_with_metadata, query)
+    locations_by_geo_level = Location
+                             .where(name: query)
+                             .pluck(:id, :geo_level)
+                             .group_by { |(_, geo_level)| geo_level }
+                             .map { |field, values| [field, values.map { |id| id[0] }] }
+                             .to_h
+
     samples = samples_with_metadata.includes(metadata: :location)
-    # Plain text locations in string_validated_value
-    samples.where(metadata: { string_validated_value: query })
-           .or(samples.where(metadata: { locations: { name: query } }))
+    # Plain text locations in string_validated_value + multi-geo-level location search
+    samples.where(
```

Comment:
you might be able to use .where().or() or inject(:or) (but probably wouldn't be an improvement)

Reply:
No, I cannot, although I tried.
It complains of an error of non-concordance between includes of `string_validated_value` and the location ones, even if you have the same includes and/or references.
Since, I did not want to dive into ActiveRecord I wrote my own clause.
This is an issue of treating locations differently... we should get rid of free string for locations.